mesos-client/src/test/scala/com/mesosphere/mesos/client/SessionActorTest.scala

```diff
@@ -0,0 +1,124 @@
+
+      Then("it sends the response to the original sender")
+      originalSender.expectMsg(httpResponse)
+    }
+
+    "make a simple call to Mesos" in withMesosStub(StatusCodes.OK) { mesos =>
+      implicit val timeout = Timeout(2.seconds)
+
+      Given("A simple Mesos API stub and a SessionActor instance")
+      val sessionActor =
+        system.actorOf(SessionActor.props(BasicAuthenticationProvider("user", "password"), mesos.createRequest))
+
+      When("we make a call through the session actor")
+      val response = (sessionActor ? Array.empty[Byte]).futureValue.asInstanceOf[HttpResponse]
+
+      Then("we receive an HTTP 200")
+      response.status should be(StatusCodes.OK)
+
+      And("Mesos was called only once")
+      Unmarshal(response.entity).to[String].futureValue.toInt should be(1)
+    }
+
+    "refresh the session token when the first call response is an HTTP 401" in withMesosStub(
+      StatusCodes.Unauthorized,
+      StatusCodes.OK) { mesos =>
+      implicit val timeout = Timeout(2.seconds)
+
+      Given("A SessionActor instance")
+      val sessionActor =
+        system.actorOf(SessionActor.props(BasicAuthenticationProvider("user", "password"), mesos.createRequest))
+
+      When("we make a call through the session actor")
+      val response = (sessionActor ? Array.empty[Byte]).futureValue.asInstanceOf[HttpResponse]
+
+      Then("we receive an HTTP 200")
+      response.status should be(StatusCodes.OK)
+
+      And("Mesos was called only once")
+      Unmarshal(response.entity).to[String].futureValue.toInt should be(2)
+    }
+  }
+
+  class Fixture() {
+
+    def requestFactory(body: Array[Byte], credentials: Option[HttpCredentials]): HttpRequest = ???
+
+  }
+
+  /**
+    * Provides a [[MesosStub]] fixture and will make sure that the stub unbinds after the test run.
+    *
+    * @param responseCodes The HTTP codes used by the stub.
+    * @param testCode The actual test definition.
+    */
+  def withMesosStub(responseCodes: StatusCode*)(testCode: MesosStub => Any): Unit = {
+    val mesos = MesosStub(responseCodes.toList)
+    try {
+      testCode(mesos)
+    } finally {
+      mesos.binding.unbind().futureValue
+    }
+  }
+
+  /**
+    * A stub for the Mesos V1 call API.
+    *
+    * All request respond with the number of calls the endpoint received. Each response takes the head of [[responseCodes]].
+    * If the list is exhausted the stub will respond with an HTTP 501 error.
+    *
+    * @param responseCodes The HTTP codes the stub will use for [[responseCodes.length]] requests.
+    */
+  case class MesosStub(var responseCodes: List[StatusCode]) {
+    import akka.http.scaladsl.server.Directives._
+
+    var calls: Int = 0
+
+    val route =
+      path("mesos") {
+        post {
+          calls += 1
+          responseCodes match {
+            case head :: tail =>
+              responseCodes = tail
```

Comment:
Here too, you're manipulating state from a different context. As an alternative make `responseCodes` a local var marked as `@volatile`